### PR TITLE
Improve @AuthenticationPrincipal meta-annotations

### PR DESCRIPTION
**Pull request by kse-music**
Closes gh-15286

### CODE REVIEW

File: messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java [hunk 1]
```diff
@@ -83,15 +94,20 @@
...
 public final class AuthenticationPrincipalArgumentResolver implements HandlerMethodArgumentResolver {
 
 	private SecurityContextHolderStrategy securityContextHolderStrategy = SecurityContextHolder
 		.getContextHolderStrategy();
 
+	private final Map<MethodParameter, Annotation> cachedAttributes = new ConcurrentHashMap<>();
+
 	private ExpressionParser parser = new SpelExpressionParser();
 
+	private PrePostTemplateDefaults defaults = new PrePostTemplateDefaults();
```
Comment (jzheaux) [inline messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java:109]: I think we'll want to introduce a new class, maybe `AuthenticationPrincipalTemplateDefaults`. By the name and JavaDoc `PrePostTemplateDefaults` is intended for the pre/post annotations.

File: messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java [hunk 2]
```diff
@@ -133,26 +149,63 @@ public void setSecurityContextHolderStrategy(SecurityContextHolderStrategy secur
...
+	/**
+	 * Configure AuthenticationPrincipal template resolution
+	 * <p>
+	 * By default, this value is <code>null</code>, which indicates that templates should
+	 * not be resolved.
+	 * @param defaults - whether to resolve AuthenticationPrincipal templates parameters
+	 * @since 6.4
+	 */
+	public void setTemplateDefaults(@Nullable PrePostTemplateDefaults defaults) {
+		if (defaults != null) {
```
Comment (jzheaux) [inline messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java:161]: Please instead don't allow null values. We don't want methods to silently not do what they say they will do.

File: messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java [hunk 3]
```diff
@@ -133,26 +149,63 @@ public void setSecurityContextHolderStrategy(SecurityContextHolderStrategy secur
...
 		}
-		Annotation[] annotationsToSearch = parameter.getParameterAnnotations();
-		for (Annotation toSearch : annotationsToSearch) {
-			annotation = AnnotationUtils.findAnnotation(toSearch.annotationType(), annotationClass);
-			if (annotation != null) {
-				return annotation;
-			}
-		}
-		return null;
+		return (T) this.cachedAttributes.computeIfAbsent(parameter, (k) -> {
```
Comment (jzheaux) [inline messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java:179]: Let's consider turning `AuthorizationAnnotationUtils` into an API that both the argument resolver and method security can use. Do you have time for that? If not, I can create a separate ticket for it.

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks for contributing this, @kse-music! I've left some initial feedback inline.

File: messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java (hunk 1 above)
Comment (kse-music) [reply messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java]: Done

File: messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java (hunk 2 above)
Comment (kse-music) [reply messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java]: Done

File: messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java (hunk 3 above)
Comment (kse-music) [reply messaging/src/main/java/org/springframework/security/messaging/context/AuthenticationPrincipalArgumentResolver.java:179]: it's my pleasure

**Comment by kse-music** [comment]
Thanks,@jzheaux ,all feedback is done

**Comment by jzheaux** [comment]
Thanks, @kse-music! Given that this needs to be placed in the other `AuthenticationPrincipalArgumentResolver`s, I extracted the logic in `AuthorizationAnnotationUtils` into an API. Please see if it works for you. If so, please also update this PR to change the other non-deprecated instances of `AuthenticationPrincipalArgumentResolver` in the project.

**Comment by kse-music** [comment]
> Thanks, @kse-music! Given that this needs to be placed in the other `AuthenticationPrincipalArgumentResolver`s, I extracted the logic in `AuthorizationAnnotationUtils` into an API. Please see if it works for you. If so, please also update this PR to change the other non-deprecated instances of `AuthenticationPrincipalArgumentResolver` in the project.

All non-deprecated instances of AuthenticationPrincipalArgumentResolver in the project has done.

**Comment by jzheaux** [comment]
Thanks, @kse-music, this is taking shape very nicely.

I believe our last step is to make this simple to configure by picking up the `AnnotationTemplateExpressionDefaults` bean in configuration.

I've added [documentation](https://github.com/spring-projects/spring-security/pull/15344/files#diff-cf8343e91909a111571bdcae3d007cb61824793545041ec143247ce3e69c1c85) so you can see the expectation that someone should be able to do:

```java
@Bean
public AnnotationTemplateExpressionDefaults templateExpressionDefaults() {
    return new AnnotationTemplateExpressionDefaults();
}
```

And now they will be able to use placeholders in any `@AuthenticationPrincipal` meta-annotation definition.

Are you able to add this to the following `spring-security-config` classes: `WebMvcSecurityConfiguration`, `ServerHttpSecurityConfiguration`, and `WebSocketMessageBrokerSecurityConfiguration`? It's a little harder, however, if you are able, it would be great to add to `WebSocketMessageBrokerSecurityBeanDefinitionParser`, too.

**Comment by kse-music** [comment]
@jzheaux ,I tried to add it

**Comment by kse-music** [comment]
@jzheaux I have added the code, please help to review it.Thank you

**Review by jzheaux — APPROVED** [review]

**Comment by jzheaux** [comment]
Thanks for all your help with this PR, @kse-music! I added a couple of commits to add more testing and also to deprecate `PrePostTemplateDefaults`. It's now merged into `main`.

**Comment by jzheaux** [comment]
If you are interested, it would be nice to do the same with `@CurrentSecurityContext`. I've added https://github.com/spring-projects/spring-security/issues/15551 to address that. That PR should go a lot more quickly.

**Comment by kse-music** [comment]
> If you are interested, it would be nice to do the same with `@CurrentSecurityContext`. I've added #15551 to address that. That PR should go a lot more quickly.

Yes,I have added PR #15553

**Pull request merged**